### PR TITLE
Filebeat - Modules Apache - Error : Fix Client IP - missing PORT

### PR DESCRIPTION
**Pull request by wixaw**
http://httpd.apache.org/docs/current/mod/core.html#errorlog

Default : ErrorLogFormat "[%t] [%l] [pid %P] %F: %E: [client %a] %M"
%a	= Client IP address and port of the request

### CODE REVIEW

File: filebeat/module/apache/error/ingest/pipeline.json [hunk 1]
```diff
@@ -5,8 +5,8 @@
       "grok": {
         "field": "message",
         "patterns": [
-          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{LOGLEVEL:log.level}\\]( \\[client %{IPORHOST:source.address}\\])? %{GREEDYDATA:message}",
-          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{DATA:apache.error.module}:%{LOGLEVEL:log.level}\\] \\[pid %{NUMBER:process.pid:long}(:tid %{NUMBER:process.thread.id:long})?\\]( \\[client %{IPORHOST:source.address}\\])? %{GREEDYDATA:message}"
+          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{LOGLEVEL:log.level}\\]( \\[client %{IPORHOST:source.address}(:%{PORT})?\\])? %{GREEDYDATA:message}",
```
Comment (jsoriano) [inline filebeat/module/apache/error/ingest/pipeline.json]: Try with this:
```suggestion
          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{LOGLEVEL:log.level}\\]( \\[client %{HOSTPORT:source.address}\\])? %{GREEDYDATA:message}",
```

File: filebeat/module/apache/error/ingest/pipeline.json [hunk 2]
```diff
@@ -5,8 +5,8 @@
...
+          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{DATA:apache.error.module}:%{LOGLEVEL:log.level}\\] \\[pid %{NUMBER:process.pid:long}(:tid %{NUMBER:process.thread.id:long})?\\]( \\[client %{IPORHOST:source.address}(:%{PORT})?\\])? %{GREEDYDATA:message}"
```
Comment (jsoriano) [inline filebeat/module/apache/error/ingest/pipeline.json]: And similarly here:
```suggestion
          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{DATA:apache.error.module}:%{LOGLEVEL:log.level}\\] \\[pid %{NUMBER:process.pid:long}(:tid %{NUMBER:process.thread.id:long})?\\]( \\[client %{HOSTPORT:source.address}\\])? %{GREEDYDATA:message}"
```

**Comment by wixaw** [comment]
Hello @jsoriano 
I had not tried it on production machines but for your request I did it.
Neither works. When I modify the pipeline.json and I restart filebeat (centos 7) nothing changes
Is this compiled somewhere?
Here is my log of example:
`[Thu Jun 27 06:58:09.169510 2019] [include:warn] [pid 15934] [client 123.123.123.123:12345] AH01374: mod_include: Options +Includes (or IncludesNoExec) wasn't set, INCLUDES filter removed: /test.html`

This is JSON view in Kibana 
```
{
  "_index": "filebeat-7.1.1-2019.06.11-000001",
  "_type": "_doc",
  "_id": "mLRMl2sBrO6GboNYYi5E",
  "_version": 1,
  "_score": null,
  "_source": {
    "agent": {
      "hostname": "servername",
      "id": "d16d6551-9c31-483a-89dc-07521dc178cc",
      "ephemeral_id": "d20240ec-96e9-4465-acf3-518ebdff12da",
      "type": "filebeat",
      "version": "7.1.1"
    },
    "process": {
      "pid": 15934
    },
    "log": {
      "file": {
        "path": "/var/log/httpd/www_error.log"
      },
      "offset": 82805,
      "level": "warn"
    },
    "message": "[client 123.123.123.123:12345] AH01374: mod_include: Options +Includes (or IncludesNoExec) wasn't set, INCLUDES filter removed: /test.html",
    "fileset": {
      "name": "error"
    },
    "input": {
      "type": "log"
    },
    "@timestamp": "2019-06-27T06:58:09.169Z",
    "apache": {
      "error": {
        "module": "include"
      }
    },
    "ecs": {
      "version": "1.0.0"
    },
    "service": {
      "type": "apache"
    },
    "host": {
      "hostname": "servername",
      "os": {
        "kernel": "3.10.0-957.1.3.el7.x86_64",
        "codename": "Core",
        "name": "CentOS Linux",
        "family": "redhat",
        "version": "7 (Core)",
        "platform": "centos"
      },
      "containerized": false,
      "name": "servername",
      "id": "98fc39052d3c4fbc9ad38e57fd127ef0",
      "architecture": "x86_64"
    },
    "event": {
      "module": "apache",
      "dataset": "apache.error"
    }
  },
  "fields": {
    "suricata.eve.timestamp": [
      "2019-06-27T06:58:09.169Z"
    ],
    "@timestamp": [
      "2019-06-27T06:58:09.169Z"
    ]
  },
  "highlight": {
    "event.module": [
      "@kibana-highlighted-field@apache@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1561618689169
  ]
}
```

We see that in message the "client" part is not extracted

File: filebeat/module/apache/error/ingest/pipeline.json [hunk 3]
```diff
@@ -5,8 +5,8 @@
       "grok": {
         "field": "message",
         "patterns": [
-          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{LOGLEVEL:log.level}\\]( \\[client %{IPORHOST:source.address}\\])? %{GREEDYDATA:message}",
-          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{DATA:apache.error.module}:%{LOGLEVEL:log.level}\\] \\[pid %{NUMBER:process.pid:long}(:tid %{NUMBER:process.thread.id:long})?\\]( \\[client %{IPORHOST:source.address}\\])? %{GREEDYDATA:message}"
+          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{LOGLEVEL:log.level}\\]( \\[client %{HOSTPORT:source.address}\\])? %{GREEDYDATA:message}",
```
Comment (jsoriano) [inline filebeat/module/apache/error/ingest/pipeline.json]: Let's go back to something more similar to your original version, but also capturing the port in the `source.port` field.
```suggestion
          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{LOGLEVEL:log.level}\\]( \\[client %{IPORHOST:source.address}(:%{POSINT:source.port})?\\])? %{GREEDYDATA:message}",
```

File: filebeat/module/apache/error/ingest/pipeline.json [hunk 4]
```diff
@@ -5,8 +5,8 @@
...
+          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{DATA:apache.error.module}:%{LOGLEVEL:log.level}\\] \\[pid %{NUMBER:process.pid:long}(:tid %{NUMBER:process.thread.id:long})?\\]( \\[client %{HOSTPORT:source.address}\\])? %{GREEDYDATA:message}"
```
Comment (jsoriano) [inline filebeat/module/apache/error/ingest/pipeline.json]: And similarly here:
```suggestion
          "\\[%{APACHE_TIME:apache.error.timestamp}\\] \\[%{DATA:apache.error.module}:%{LOGLEVEL:log.level}\\] \\[pid %{NUMBER:process.pid:long}(:tid %{NUMBER:process.thread.id:long})?\\]( \\[client %{IPORHOST:source.address}(:%{POSINT:source.port})?\\])? %{GREEDYDATA:message}"
```

File: filebeat/module/apache/error/test/test.log [hunk 5]
```diff
@@ -1,3 +1,4 @@
 [Mon Dec 26 16:22:08 2016] [error] [client 192.168.33.1] File does not exist: /var/www/favicon.ico
 [Mon Dec 26 16:15:55.103786 2016] [core:notice] [pid 11379] AH00094: Command line: '/usr/local/Cellar/httpd24/2.4.23_2/bin/httpd'
 [Fri Sep 09 10:42:29.902022 2011] [core:error] [pid 35708:tid 4328636416] [client 72.15.99.187] File does not exist: /usr/local/apache2/htdocs/favicon.ico
+[Thu Jun 27 06:58:09.169510 2019] [include:warn] [pid 15934] [client 123.123.123.123:12345] AH01374: mod_include: Options +Includes (or IncludesNoExec) wasn't set, INCLUDES filter removed: /test.html
```
Comment (jsoriano) [inline filebeat/module/apache/error/test/test.log]: We will also need to add the expected result of this entry to the `test.log-expected.json` file. You can automatically generate it running the tests with `GENERATE=1` from the filebeat directory:
```
GENERATE=1 INTEGRATION_TESTS=1 BEAT_STRICT_PERMS=false TESTING_FILEBEAT_MODULES=apache nosetests tests/system/test_modules.py
```

As doing the setup for these tests is a bit time-consuming sometimes, I have generated it:
```
    {
        "@timestamp": "2019-06-27T06:58:09.169Z",
        "apache.error.module": "include",
        "event.dataset": "apache.error",
        "event.module": "apache",
        "fileset.name": "error",
        "input.type": "log",
        "log.level": "warn",
        "log.offset": 384,
        "message": "AH01374: mod_include: Options +Includes (or IncludesNoExec) wasn't set, INCLUDES filter removed: /test.html",
        "process.pid": 15934,
        "service.type": "apache",
        "source.address": "123.123.123.123",
        "source.geo.city_name": "Beijing",
        "source.geo.continent_name": "Asia",
        "source.geo.country_iso_code": "CN",
        "source.geo.location.lat": 39.9288,
        "source.geo.location.lon": 116.3889,
        "source.geo.region_iso_code": "CN-BJ",
        "source.geo.region_name": "Beijing",
        "source.ip": "123.123.123.123",
        "source.port": "12345"
    }
```

**Comment by wixaw** [comment]
Yesss
Perfect it's work !!
Thanks

```
{
  "_index": "filebeat-7.1.1-2019.06.11-000001",
  "_type": "_doc",
  "_id": "iLivmGsBrO6GboNYAmxo",
  "_version": 1,
  "_score": null,
  "_source": {
    "agent": {
      "hostname": "servername",
      "id": "d16d6551-9c31-483a-89dc-07521dc178cc",
      "ephemeral_id": "5a212816-3cd7-4378-bfbd-79afc938de17",
      "type": "filebeat",
      "version": "7.1.1"
    },
    "process": {
      "pid": 24096
    },
    "log": {
      "file": {
        "path": "/var/log/httpd/www_error.log"
      },
      "offset": 91097,
      "level": "warn"
    },
    "source": {
      "geo": {
        "continent_name": "Europe",
        "country_iso_code": "FR",
        "location": {
          "lon": 2.xxxxxxxx,
          "lat": 4x.xxxxx
        }
      },
      "address": "123.123.123.123,
      "port": "46142",
      "ip": "123.123.123.111"
    },
    "message": "AH01374: mod_include: Options +Includes (or IncludesNoExec) wasn't set, INCLUDES filter removed: /test.html",
    "fileset": {
      "name": "error"
    },
    "input": {
      "type": "log"
    },
    "@timestamp": "2019-06-27T13:25:29.234Z",
    "apache": {
      "error": {
        "module": "include"
      }
    },
    "ecs": {
      "version": "1.0.0"
    },
    "service": {
      "type": "apache"
    },
    "host": {
      "hostname": "servername",
      "os": {
        "kernel": "3.10.0-957.1.3.el7.x86_64",
        "codename": "Core",
        "name": "CentOS Linux",
        "family": "redhat",
        "version": "7 (Core)",
        "platform": "centos"
      },
      "containerized": false,
      "name": "servername",
      "id": "98fc39052d3c4fbc9ad38e57fd127ef0",
      "architecture": "x86_64"
    },
    "event": {
      "module": "apache",
      "dataset": "apache.error"
    }
  },
  "fields": {
    "suricata.eve.timestamp": [
      "2019-06-27T13:25:29.234Z"
    ],
    "@timestamp": [
      "2019-06-27T13:25:29.234Z"
    ]
  },
  "highlight": {
    "event.module": [
      "@kibana-highlighted-field@apache@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1561641929234
  ]
}
```

**Comment by jsoriano** [comment]
jenkins, test this

**Comment by wixaw** [comment]
It's signed :)

**Comment by jsoriano** [comment]
@wixaw sorry, one last thing, could you add a changelog entry to `CHANGELOG.next.asciidoc`? Thanks!

**Comment by jsoriano** [comment]
jenkins, test this again please

**Comment by jsoriano** [comment]
@wixaw merged, thanks!

**Pull request merged**